deepinv/loss/score.py

```diff
@@ -1,41 +1,161 @@
+
+    .. warning::
+
+        The user should provide a backbone model :math:`S`
+        to :meth:`adapt_model` which returns the full reconstruction network
+        :math:`R`, which is mandatory to compute the loss properly.
+
+    .. warning::
+
+        This class uses the inference formula for the Poisson noise case
+        which differs from the one proposed in Noise2Score.
+
+    .. note::
+
+        This class does not support general inverse problems, it is only designed for denoising problems.
+
+    :param None, torch.nn.Module noise_model: Noise distribution corrupting the measurements
+        (see :ref:`the physics docs <physics>`). Options are :class:`deepinv.physics.GaussianNoise`,
+        :class:`deepinv.physics.PoissonNoise`, :class:`deepinv.physics.GammaNoise` and
+        :class:`deepinv.physics.UniformGaussianNoise`. By default, it uses the noise model associated with
+        the physics operator provided in the forward method.
+    :param int total_batches: Total number of training batches (epochs * number of batches per epoch).
+    :param tuple delta: Tuple of two floats representing the minimum and maximum noise level,
+        which are annealed during training.
+
+    |sep|
+
+
+    :Example:
+
+        >>> import torch
+        >>> import deepinv as dinv
+        >>> sigma = 0.1
+        >>> physics = dinv.physics.Denoising(dinv.physics.GaussianNoise(sigma))
+        >>> model = dinv.models.DnCNN(depth=2, pretrained=None)
+        >>> loss = dinv.loss.ScoreLoss(total_batches=1, delta=(0.001, 0.1))
+        >>> model = loss.adapt_model(model) # important step!
+        >>> x = torch.ones((1, 3, 5, 5))
+        >>> y = physics(x)
+        >>> x_net = model(y, physics, update_parameters=True) # save score loss in forward
+        >>> l = loss(model)
+        >>> print(l.item() > 0)
+        True
     """
 
-    def __init__(self, delta):
+    def __init__(self, noise_model=None, total_batches=1000, delta=(0.001, 0.1)):
         super(ScoreLoss, self).__init__()
-        self.name = "score"
-        self.metric = torch.nn.MSELoss()
+        self.total_batches = total_batches
         self.delta = delta
+        self.noise_model = noise_model
 
-    def forward(self, y, model, **kwargs):
+    def forward(self, model, **kwargs):
         r"""
-        Computes the Score loss.
+        Computes the Score Loss.
 
-        :param torch.Tensor y: measurements.
-        :param torch.nn.Module model: Reconstruction function.
-        :return: (torch.Tensor) loss.
+        :param torch.Tensor y: Measurements.
+        :param deepinv.physics.Physics physics: Forward operator associated with the measurements.
+        :param torch.nn.Module model: Reconstruction model.
+        :return: (torch.Tensor) Score loss.
         """
-        std = np.randn() * self.delta
-        noise = torch.randn_like(y)
-        return self.metric(noise, std * model(std * noise + y))
+        return model.get_error()
```

Comment:
Can we add a test to ensure that we indeed computer the error beforehand ?

Reply:
do you mean a pytest? or a flag that makes sure that the error has been recomputed?